piptools/repositories/pypi.py

```diff
@@ -434,22 +424,23 @@ def _wheel_support_index_min(self: Wheel, tags: List[Tag]) -> int:
 
         original_wheel_supported = Wheel.supported
         original_support_index_min = Wheel.support_index_min
-        original_cache = self._available_candidates_cache
+        original_candidates_cache = self.finder.find_all_candidates
 
         Wheel.supported = _wheel_supported
         Wheel.support_index_min = _wheel_support_index_min
-        self._available_candidates_cache = {}
-
         # If we don't clear this cache then it can contain results from an
         # earlier call when allow_all_wheels wasn't active. See GH-1532
-        self.finder.find_all_candidates.cache_clear()
+        # TODO: Python >= 3.9: Use `lru_cache().cache_parameters()`.
+        self.finder.find_all_candidates = lru_cache(
+            maxsize=self.finder.find_all_candidates.cache_info().maxsize
+        )(self.finder.find_all_candidates.__wrapped__).__get__(self.finder)
```

Comment:
Could you elaborate on this block? This part looks tricky.

Reply:
Sure!

`self.finder.find_all_candidates` is an `lru_cached` instance method with signature `(req_name: str) -> ...`. The function that the the cache wraps, i.e. `self.finder.find_all_candidates.__wrapped__` (https://github.com/pypa/pip/blob/ca2d9f41931a449b8c1b27d02031199d91af93e7/src/pip/_internal/index/package_finder.py#L810-L860; note that L809 is not included in this highlight) has signature `(self: PackageFinder, req_name: str) -> ...`. Note the additional `self` there—the L810–L860 function that `@lru_cache(maxsize=None)` wraps has `self` in its signature still.<sup>1</sup>

`x = lru_cache(maxsize=self.finder.find_all_candidates.cache_info().maxsize)(self.finder.find_all_candidates.__wrapped__)` creates a new `lru_cache`d version of the same underlying wrapped function (L810–L860) with the same caching parameters, so that we have a fresh cache to use while the context manager `allow_all_wheels` is open. Note that `x` has the same signature as the wrapped function, i.e. it still needs `self` as its first argument. If we just did `finder.x = x`, we'd have to call it as `finder.x(finder, "foo")`, since it still needs `self` as an argument. To actually bind `x` as a method of `finder`, we need to invoke it as a descriptor `finder.x = x.__get__(finder)`.

If you prefer, `finder.x = types.MethodType(finder, x)` would be equivalent to `finder.x = x.__get__(finder)` here.

[Descriptors](https://docs.python.org/3/glossary.html#term-descriptor) can definitely be a bit tricky, but they're pretty core to the language. https://docs.python.org/3/howto/descriptor.html#simple-example-a-descriptor-that-returns-a-constant and https://docs.python.org/3/howto/descriptor.html#functions-and-methods definitely explains `__get__` descriptors better than I can if you haven't seen them (explicitly) before. (You have _definitely_ seen them implicitly though–every instance/class method is the result of a `__get__` binding!!)

<sup>1</sup>: <details>
<summary>Click to expand footnote</summary>
  
Aside: `lru_cache`ing instance/class methods can cause garbage collector deadlocks if one is not careful! E.g.

```
class Foo:
   @lru_cache(maxsize=None)
   def bar(self, x: int): ...
```

This creates a single cache, not a separate cache for each instance of `Foo`! The single cache is keyed on `(self, x)`, though, not just on `(x,)`.

Doing

```
foo = Foo()
print(foo)  # repr(foo) == <Foo object at 0x1234>
foo.bar(0)
```

will add new entry to the cache, keyed on `(foo, 0)`. Note that the instance `foo` is part of the new cache key. Now suppose we either do `foo = something_else`, or we leave the context of the variable `foo`. Both of these would *usually* make `<Foo object at 0x1234>` ready to be garbage collected. Here, however, our cache is still keeping around a reference to `Foo` and prevent garbage collection. This is a memory leak if `maxsize=None`!
</details>